File: annotations/src/main/java/com/siimkinks/sqlitemagic/annotation/Table.java
package com.siimkinks.sqlitemagic.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks class to be a table in SQLite database.<br>
 * <p>
 * If annotated class is
 * <p>
 * <b>Value object:</b><br>
 * {@link Column} annotations must be on methods<br>
 * Must have {@link Id} annotated method that returns long<br>
 * </p>
 * <p>
 * <b>Regular object:</b><br>
 * {@link Column} annotations must be on fields<br>
 * Optionally can have {@link Id} annotated long field. If missing, auto-incremented
 * id column with "_id" name is created along with package private long id field in
 * annotated model<br>
 * </p>
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface Table {
  /**
   * Table name.<br>
   * Defaults to lower cased class name.
   *
   * @return Table name
   */
  String value() default "";

  /**
   * Persist all possible fields in object.<br>
   * If true system tries to create sql table columns
   * from all fields using all the default behavior of {@link Column}.
   * If any field has {@link Column} annotation then it uses its preferences.
   * If any field is annotated with {@link IgnoreColumn} then this field is ignored.
   *
   * @return True if all fields should be persisted, false if not.
   */
  boolean persistAll() default false;

  /**
   * Respect access methods when accessing model fields.<br>
   * Access methods can be with names representing field names e.g.
   * {@code name(), name(String)} or with JavaBeans-style prefixes like
   * {@code getName(), setName(String), isMale()}.<br>
   * This parameter is only applicable on non-value objects.
   *
   * @return True if system should access model fields with access methods, false if not.
   */
  boolean useAccessMethods() default false;
}

File: annotations/src/main/kotlin/com/siimkinks/sqlitemagic/annotation/Table.kt
package com.siimkinks.sqlitemagic.annotation

import kotlin.annotation.AnnotationRetention.BINARY
import kotlin.annotation.AnnotationTarget.CLASS

/**
 * Marks a class as a table in the SQLite database.
 *
 * Supported table models are constructor-backed classes whose persisted properties correspond
 * exactly to an accessible primary constructor, and mutable classes that generated Kotlin can
 * instantiate without arguments and access through readable and writable properties.
 * Constructor-backed classes do not need to be data classes. The declaration's source language is
 * irrelevant when it exposes the same construction and property surface to Kotlin.
 *
 * An [Id] column is optional. When present, it may use any otherwise supported persisted type.
 * When absent, no implicit ID column is generated.
 *
 * A table without an ID supports insert operations and explicit update or delete statements whose
 * affected rows are selected by the caller. When it has a non-null single-column [Unique]
 * property, identity-dependent entity operations are also available but require the generated
 * unique-column property as an argument to `execute` or `observe`. A nullable unique column or
 * composite unique index cannot serve as that operation key. Single-entity persist reports
 * whether it inserted a row, including SQLite's generated row ID, or updated an existing row.
 *
 * @property value Table name. Defaults to the lower-cased class name with camel case replaced by `_`.
 *
 * @property persistAll Whether all possible properties should be persisted as columns using
 *     [Column] defaults. A property's [Column] annotation overrides those defaults, while
 *     [IgnoreColumn] excludes the property.
 *
 * @property options SQLite table definition options. See [TableOption].
 */
@Target(CLASS)
@Retention(BINARY)
annotation class Table(
  val value: String = "",
  val persistAll: Boolean = true,
  val options: Array<TableOption> = []
)
